contracts/ERC7432/NftRolesRegistryVault.sol

```diff
@@ -0,0 +1,236 @@
+
+        delete originalOwners[_tokenAddress][_tokenId];
+        delete roles[_tokenAddress][_tokenId][_roleId];
+        delete latestRecipients[_tokenAddress][_tokenId][_roleId];
+        emit RoleRevoked(_tokenAddress, _tokenId, _roleId);
+    }
+
+    function setRoleApprovalForAll(address _tokenAddress, address _operator, bool _approved) external override {
+        tokenApprovals[msg.sender][_tokenAddress][_operator] = _approved;
+        emit RoleApprovalForAll(_tokenAddress, _operator, _approved);
+    }
+
+    /** ERC-7432 View Functions **/
+
+    function recipientOf(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (address recipient_) {
+        if (
+            _isTokenDeposited(_tokenAddress, _tokenId) &&
+            roles[_tokenAddress][_tokenId][_roleId].expirationDate > block.timestamp
+        ) {
+            return roles[_tokenAddress][_tokenId][_roleId].recipient;
+        }
+        return address(0);
+    }
+
+    function roleData(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (bytes memory data_) {
+        if (!_isTokenDeposited(_tokenAddress, _tokenId)) {
+            return '';
+        }
+        return roles[_tokenAddress][_tokenId][_roleId].data;
+    }
+
+    function roleExpirationDate(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (uint64 expirationDate_) {
+        if (!_isTokenDeposited(_tokenAddress, _tokenId)) {
+            return 0;
+        }
+        return roles[_tokenAddress][_tokenId][_roleId].expirationDate;
+    }
+
+    function isRoleRevocable(
+        address _tokenAddress,
+        uint256 _tokenId,
+        bytes32 _roleId
+    ) external view returns (bool revocable_) {
+        if (!_isTokenDeposited(_tokenAddress, _tokenId)) {
+            return false;
+        }
+        return roles[_tokenAddress][_tokenId][_roleId].revocable;
+    }
+
+    function isRoleApprovedForAll(
+        address _tokenAddress,
+        address _owner,
+        address _operator
+    ) public view returns (bool) {
+        return tokenApprovals[_owner][_tokenAddress][_operator];
+    }
+
+    /** ERC-7432 Vault Extension Functions **/
+
+    function withdraw(address _tokenAddress, uint256 _tokenId) external override {
+        address originalOwner = originalOwners[_tokenAddress][_tokenId];
+
+        require(_isWithdrawable(_tokenAddress, _tokenId), 'NftRolesRegistryVault: NFT is not withdrawable');
+
+        require(
+            originalOwner == msg.sender || isRoleApprovedForAll(_tokenAddress, originalOwner, msg.sender),
+            'NftRolesRegistryVault: sender must be owner or approved'
+        );
+
+        IERC721(_tokenAddress).transferFrom(address(this), originalOwner, _tokenId);
+        delete originalOwners[_tokenAddress][_tokenId];
+        emit Withdraw(originalOwner, _tokenAddress, _tokenId);
```

Comment:
place transfer after storage deletion to follow Checks Effects Interactions Pattern

```suggestion
        delete originalOwners[_tokenAddress][_tokenId];
        emit Withdraw(originalOwner, _tokenAddress, _tokenId);
        IERC721(_tokenAddress).transferFrom(address(this), originalOwner, _tokenId);
```

Reply:
Done